### DISCUSSION TITLE
Decentralisation

### DISCUSSION THREAD

**Discussion by Nintorac** [NVIDIA/NVFlare#300] [Ideas]
In a federated learning system you have a set of actors that belong to different parties of the federation and 

These actors either have data, or control workflows. 

For example;
institution `A` has two researchers `a_r0`, `a_r1` and three private data sources `a_d0`, `a_d1`, `a_d2`
institution `B` has three researchers `b_r0`, `b_r1`, and `b_r2` and two private data sources `b_d0`, `b_d1`

In this scenario there is a bipartite matching where one side represents researchers and the other data sources.

Since we are implementing a messaging system there is no need to have a central "Server" authority. We just need some way to negotiate role and permissions in order to manage who can send messages where.

**Comment by yanchengnv** [comment]
This is a good topic. We realize that the essence of NVFLARE is two things: workers that can do things, and a control logic that directs/coordinates the workers. These two concepts don't have to be mapped to a centralized server/client topology.

We chose to use this centralized topology for practical reasons:
- most easily understood
- easy to implement/deploy, especially in industries like healthcare and finance, where institutions have very strict policy (e.g. opening port is strictly disallowed)
- easy to write the control logic

Keep in mind that control logic is to be written by researchers, who prefer simple procedural programming.

**Reply by Nintorac** [reply]
From my perspective there are only actors. Actors can be either service providers or service consumers.

The control logic then comes directly from the service consumer in their script.

Worth noting that it would be useful to allow service providers to also consume from other service providers.

> Keep in mind that control logic is to be written by researchers, who prefer simple procedural programming.

All of this should be totally abstracted away from the researcher, they do not care how their requests are fulfilled

> easy to implement/deploy, especially in industries like healthcare and finance, where institutions have very strict policy (e.g. opening port is strictly disallowed)

This is important, the way I see it being solved in this hypothetical decentralized federated learning env is by adding another actor that does the job of message routing. The `router actor` is then given a globally routable (i.e routable by all actors) address to which service consumers can connect to in order to have messages relayed between secure environments.

**Comment by poa010101** [comment]
using Blockchain and smart contractor?
